iceoryx_posh/include/iceoryx_posh/experimental/popo/base_subscriber.hpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_EXPERIMENTAL_POSH_POPO_BASE_SUBSCRIBER_HPP
+#define IOX_EXPERIMENTAL_POSH_POPO_BASE_SUBSCRIBER_HPP
+
+#include "iceoryx_posh/mepoo/chunk_header.hpp"
+#include "iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp"
+#include "iceoryx_posh/popo/condition.hpp"
+#include "iceoryx_utils/cxx/expected.hpp"
+#include "iceoryx_utils/cxx/optional.hpp"
+#include "iceoryx_utils/cxx/unique_ptr.hpp"
+
+namespace iox {
+namespace popo {
+
+using uid_t = uint64_t;
+
+enum class SubscriberError : uint8_t
+{
+    UNKNOWN
+};
+
+template<typename T, typename port_t = popo::SubscriberPortUser>
+class BaseSubscriber : public Condition
+{
+public:
+
+    BaseSubscriber(const BaseSubscriber& other) = delete;
+    BaseSubscriber& operator=(const BaseSubscriber&) = delete;
+    BaseSubscriber(BaseSubscriber&& rhs) = delete;
+    BaseSubscriber& operator=(BaseSubscriber&& rhs) = delete;
+    ~BaseSubscriber() = default;
+
+    ///
+    /// @brief uid Get the unique ID of the subscriber.
+    /// @return The subscriber's unique ID.
+    ///
+    uid_t uid() const noexcept;
+
+    ///
+    /// @brief getServiceDescription Get the service description of the subscriber.
+    /// @return The service description.
+    ///
+    capro::ServiceDescription getServiceDescription() const noexcept;
+
+    ///
+    /// @brief subscribe Initiate subscription.
+    /// @param cacheSize
+    /// @return
+    ///
+    cxx::expected<SubscriberError> subscribe(const uint64_t queueCapacity = MAX_SUBSCRIBER_QUEUE_CAPACITY) noexcept;
+
+    ///
+    /// @brief getSubscriptionState Get current subscription state.
+    /// @return The current subscription state.
+    ///
+    SubscribeState getSubscriptionState() const noexcept;
+
+    ///
+    /// @brief unsubscribe Unsubscribes if currently subscribed, otherwise do nothing.
+    ///
+    void unsubscribe() noexcept;
+
+    ///
+    /// @brief hasData Check if sample is available.
+    /// @return True if a new sample is available.
+    ///
+    bool hasData() const noexcept;
+
+    ///
+    /// @brief receive Receive the next sample if available.
+    /// @return
+    /// @details Sample is automatically released when it goes out of scope.
+    ///
+    cxx::optional<cxx::unique_ptr<T>> receive() noexcept;
```

Comment:
somehow strange that on publisher side you publish a sample and on subscriber side you don't receive a sample but a unique_ptr<T>. In ara:.com you have sample on both sides with different sample smart pointers

Reply:
I changed this this morning, take a look.